### PR TITLE
Merge GL codebases

### PR DESCRIPTION
**Pull request by Beyley**
Most of the OpenGL code now resides in the OpenGL folder, with a light abstraction layer providing a single API for talking with OpenGL, OpenGLES, and Legacy OpenGL using only modern OpenGL
This also fixes a bug in the GL20 backend which would cause the first quad of multiple batches to be dropped

### CODE REVIEW

**Review by Eeveelution — APPROVED** [review]

**Comment by Eeveelution** [comment]
I think the way it's being done is good, and generalizing the GL backends is also pretty necessary since there's a fuckton of Code Duplication thanks to us supporting 3 GL backends.

**Comment by Beyley** [comment]
> I think the way it's being done is good, and generalizing the GL backends is also pretty necessary since there's a fuckton of Code Duplication thanks to us supporting 3 GL backends.

we might be able to just completely ditch most of our GLES code and switch the quad renderer for 4.1 and ES to the same code, which would be nice

**Comment by Eeveelution** [comment]
that would be possible yeah

**Comment by Beyley** [comment]
Havent moved this out to another project (im very lazy rn and that can be done later)

File: Furball.Vixie/Graphics/Backends/OpenGL/IGLBasedBackend.cs [hunk 1]
```diff
@@ -14,12 +15,57 @@ public interface IGLBasedBackend {
...
+        public        void   Uniform1(int            getUniformLocation, int    f);
+        public        void   Uniform2(int            getUniformLocation, float  f, float f2);
+        public        void   Uniform2(int            getUniformLocation, int    f, int   f2);
+        public        void   DeleteProgram(uint      programId);
+        public        uint   GenVertexArray();
+        public        void   EnableVertexAttribArray(uint u);
+        public unsafe void   VertexAttribPointer(uint     u, int currentElementCount, VertexAttribPointerType currentElementType, bool currentElementNormalized, uint  getStride, void* offset);
+        public unsafe void   VertexAttribIPointer(uint    u, int currentElementCount, VertexAttribIType       vertexAttribIType,  uint getStride,                void* offset);
+        public        void   BindVertexArray(uint         arrayId);
+        public        void   DeleteVertexArray(uint       arrayId);
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL/IGLBasedBackend.cs]: the automatic snapping thingy is making this look horrifying

File: Furball.Vixie/Graphics/Backends/OpenGL20/LineRendererGL20.cs [hunk 2]
```diff
@@ -14,9 +15,9 @@ private struct LineData {
         
         private readonly OpenGL20Backend _backend;
 
-        private const int BATCH_MAX = 128; //cut this in two for actual line could, as we use 2 LineData structs per line :^)
+        private const int BATCH_MAX = 128; //cut this in two for actual line couyt, as we use 2 LineData structs per line :^)
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/LineRendererGL20.cs]: why did you fuck up perfectly good spelling

File: Furball.Vixie/Graphics/Backends/OpenGL20/OpenGL20Backend.cs [hunk 3]
```diff
@@ -176,5 +179,122 @@ public class OpenGL20Backend : GraphicsBackend, IGLBasedBackend {
...
+        }
+        public unsafe void VertexAttribPointer(uint u, int currentElementCount, Silk.NET.OpenGL.VertexAttribPointerType currentElementType, bool currentElementNormalized, uint getStride, void* offset) {
+            this.gl.VertexAttribPointer(u, currentElementCount, (GLEnum)currentElementType, currentElementNormalized, getStride, offset);
+        }
+        public unsafe void VertexAttribIPointer(uint u, int currentElementCount, Silk.NET.OpenGL.VertexAttribIType vertexAttribIType, uint getStride, void* offset) {
+            this.gl.VertexAttribIPointer(u, currentElementCount, (GLEnum)vertexAttribIType, getStride, offset);
+        }
+        public void BindVertexArray(uint arrayId) {
+            this.gl.BindVertexArray(arrayId);
+        }
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL20/OpenGL20Backend.cs:347]: at the VERY least seperate each function with 1 new line, this is all crammed together so tightly so thats the VERY least you can do

File: Furball.Vixie/Graphics/Backends/OpenGL41/OpenGL41Backend.cs [hunk 4]
```diff
@@ -270,5 +269,121 @@ public class OpenGL41Backend : GraphicsBackend, IGLBasedBackend {
...
+        public unsafe void VertexAttribPointer(uint u, int currentElementCount, Silk.NET.OpenGL.VertexAttribPointerType currentElementType, bool currentElementNormalized, uint getStride, void* offset) {
+            this.gl.VertexAttribPointer(u, currentElementCount, (GLEnum)currentElementType, currentElementNormalized, getStride, offset);
+        }
+        public unsafe void VertexAttribIPointer(uint u, int currentElementCount, Silk.NET.OpenGL.VertexAttribIType vertexAttribIType, uint getStride, void* offset) {
+            this.gl.VertexAttribIPointer(u, currentElementCount, (GLEnum)vertexAttribIType, getStride, offset);
+
+        }
+        public void BindVertexArray(uint arrayId) {
+            this.gl.BindVertexArray(arrayId);
+        }
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGL41/OpenGL41Backend.cs:435]: same here with the 1 line spacing

File: Furball.Vixie/Graphics/Backends/OpenGLES/OpenGLESBackend.cs [hunk 5]
```diff
@@ -263,5 +273,123 @@ public class OpenGLESBackend : GraphicsBackend, IGLBasedBackend {
...
+        public unsafe void VertexAttribPointer(uint u, int currentElementCount, VertexAttribPointerType currentElementType, bool currentElementNormalized, uint getStride, void* offset) {
+            this.gl.VertexAttribPointer(u, currentElementCount, (GLEnum)currentElementType, currentElementNormalized, getStride, offset);
+        }
+        public unsafe void VertexAttribIPointer(uint u, int currentElementCount, VertexAttribIType vertexAttribIType, uint getStride, void* offset) {
+            this.gl.VertexAttribIPointer(u, currentElementCount, (GLEnum)vertexAttribIType, getStride, offset);
+
+        }
+        public void BindVertexArray(uint arrayId) {
+            this.gl.BindVertexArray(arrayId);
+        }
```
Comment (Eeveelution) [inline Furball.Vixie/Graphics/Backends/OpenGLES/OpenGLESBackend.cs:441]: same here

**Review by Eeveelution — CHANGES_REQUESTED** [review]

File: Furball.Vixie/Graphics/Backends/OpenGL20/LineRendererGL20.cs (hunk 2 above)
Comment (Beyley) [reply Furball.Vixie/Graphics/Backends/OpenGL20/LineRendererGL20.cs]: i-- dont... know

Comment (Beyley) [reply Furball.Vixie/Graphics/Backends/OpenGL20/LineRendererGL20.cs]: OH i know why, i tried to correct it from could -> count, but fucked up

**Review by Eeveelution — APPROVED** [review]
looks fine by me

**Pull request merged**